### DISCUSSION TITLE
Could not load from page that was just created.

### DISCUSSION THREAD

**Discussion by Yari27** [cefsharp/CefSharp#4551] [General]
### Is there an existing issue for this?

- [X] I have searched both open/closed issues, no issue already exists.

### CefSharp Version

114.2.120

### Operating System

Windows 10

### Architecture

x64

### .Net Version

.Net.4.8

### Implementation

WPF

### Reproduction Steps

```
private string address;
private string filePath = @"C:\Users\user\OneDrive\Documents\index.html";

public string Address
        {
            get => address;
            set
            {
                if (value != address)
                {
                    address = value;
                    OnPropertyChanged();
                }
            }
        }

//First try - File not exist.
Address = $"file:///{filePath}"; 

//Second try - The file has been created
OnPropertyChanged(nameof(Address));
```

### Expected behavior

- First attempt to load page from file (file does not exist)
nothing loads.
- Second attempt to load the page (file already exists)
the page has been loaded

### Actual behavior

- First attempt to load page from file (file does not exist)
nothing loads.
- Second attempt to load the page (file already exists)
does not load the page from the given file. (Page source is empty ```<html><head></head><body></body></html>```)

### Does this problem also occur in the CEF Sample Application

Not Tested

**Comment by amaitland** [comment]
Converted to discussion as this very unlikely a bug in `CefSharp`.

Assuming you've bound to the `Address` [DependencyProperty](https://learn.microsoft.com/en-us/dotnet/api/system.windows.dependencyproperty?view=windowsdesktop-7.0) then it's likely the behaviour you are seeing is expected.

Your firing the property change nofiticaiton without changing the actual value. From memory `WPF` ignores property change notifications when `newValue` and `oldValue` are equal. A quick look at the source yields.

> // Change notifications are fired when the value actually changed

https://referencesource.microsoft.com/#WindowsBase/Base/System/Windows/DependencyObject.cs,1551

Your call is basically ignored by `WPF`, which is expected as you didn't change the value.

**Comment by Yari27** [comment]
Then why when I change the content of the page in the file from time to time, it refreshes, even though the address is the same all the time?

When the starting address is empty or the file does not exist, the page will never be displayed on subsequent refreshes.
But when the starting address is, for example, "www.google.com or an existing file, the page will always refresh when I call `OnPropertyChanged(nameof(Address))` even for the same address;

**Reply by amaitland** [reply]
> even though the address is the same all the time?

The `Address`  (`DependencyProperty`) is updated to reflect the browsers internal changes of url. 

https://github.com/cefsharp/CefSharp/blob/cefsharp/114/CefSharp.Wpf/ChromiumWebBrowser.cs#L1126

> When the starting address is empty or the file does not exist, the page will never be displayed on subsequent refreshes.

- `WPF` should only raise a change notification if the values are different
- Depending on how your binding is configured you will get different results.

Calling `OnPropertyChanged(nameof(Address))` is not the equivalent of calling `Reload`.


You can add a breakpoint in the changed handler to get a better understanding of what's going on.

https://github.com/cefsharp/CefSharp/blob/cefsharp/114/CefSharp.Wpf/ChromiumWebBrowser.cs#L1254-L1277

**Reply by Yari27** [reply]
I have an MVVM application and I need 100% certainty/efficiency that the page will be refreshed every time for the same address, what property will give me that?

**Comment by amaitland** [comment]
> what property will give me that?

`Reload` is fundamentally a method, it doesn't map to a property directly. There is a [ReloadCommand](https://github.com/cefsharp/CefSharp/blob/cefsharp/114/CefSharp.Wpf/IWpfWebBrowser.cs#L35-L39) available for use in the view.

> I have an MVVM application

Fundamentally you can call [ChromiumWebBrowser.Reload](https://cefsharp.github.io/api/112.2.x/html/M_CefSharp_WebBrowserExtensions_Reload_1.htm)

How you do this is really up to you and how your application is architected.
- Send a message from the ViewModel to the View to reload the browser
- Bind the [WebBrowser](https://github.com/cefsharp/CefSharp/blob/cefsharp/114/CefSharp.Wpf.Example/Views/BrowserTabView.xaml#L441) to your `ViewModel` and access the browser via it's interface.
- etc

If you are unsure then see what the recommended approach is for your `MVVM framework`.

**Reply by Yari27** [reply]
Thanks, it works great now, just what I was looking for.

P.S. The question is, if the Address is empty or null, the ReloadCommand should be active? What is Reloading?

**Reply by amaitland** [reply]
You can look at the code to see the current behaviour. It's not something I've reviewed in a long time.

If you think there are improvements that would make sense to all users (not just your own scenario) then by all means submit a `PR`.
